vpa-recommender: Add support for configuring global max allowed resources

ialidzhikov opened the pull request:
#### What type of PR is this?

/kind feature

#### What this PR does / why we need it:
See  https://github.com/kubernetes/autoscaler/issues/7147

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes https://github.com/kubernetes/autoscaler/issues/7147

#### Special notes for your reviewer:
N/A

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
vpa-recommender now supports two new flags - `--container-recommendation-max-allowed-cpu` and `--container-recommendation-max-allowed-memory`. The flags make possible to configure the global max allowed resources the vpa-recommender can recommend for a container. The flags aim to address a known limitation of VPA (vpa-recommender, in particular) that it is not aware of the cluster's capacity and can make Pod unschedulable after applying a recommendation that cannot fit in the cluster (recommendation higher than largest Node's allocatable).
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

adrianmoisey commented:
Sorry, but I've just gotten a PR merged (https://github.com/kubernetes/autoscaler/pull/7548) that will merge conflict with yours.
It's only documentation related.

adrianmoisey commented on vertical-pod-autoscaler/pkg/recommender/main.go:
```diff
@@ -116,6 +119,9 @@ const (
 )
 
 func main() {
+	flag.Var(&maxAllowedCPU, "max-allowed-cpu", "Maximum amount of CPU that will be recommended for a container. VerticalPodAutoscaler-level maximum allowed takes precedence over the global maximum allowed.")
+	flag.Var(&maxAllowedMemory, "max-allowed-memory", "Maximum amount of memory that will be recommended for a container. VerticalPodAutoscaler-level maximum allowed takes precedence over the global maximum allowed.")
```
Should these flags live in `pkg/recommender/logic/recommender.go`, alongside `pod-recommendation-min-memory-mb` and `pod-recommendation-min-cpu-millicores` ?

Also, should they have similar names to the min flags, such as: `container-recommendation-max-memory-mb` and `container-recommendation-max-cpu-millicores` ?

ialidzhikov replied:
In https://github.com/kubernetes/autoscaler/issues/7147#issuecomment-2501042262, I already tried to explain why  Pod-level flags like `--container-recommendation-max-memory-mb`/`--container-recommendation-max-cpu-millicores` don't make sense for max allowed:
> On the other side, recommendations in VPA are on container-level. VPA can be enabled for > 1 container in a Pod. The question is how VPA is expected to distribute the Pod-level max allowed recommendation across the many containers in the Pod. Some approaches are - distribute the configured max allowed recommendation values evenly; distribute proportionally (proportionally based on what?). My conclusion is that VPA should not do that as it does not know about the workload.

IMO the existing `--container-recommendation-max-memory-mb`/`--container-recommendation-max-cpu-millicores` are not okay. They are on Pod-level and the vpa-recommender splits them evenly to the number of containers in the Pod. The implementation of these flags is such that it overwrites the `uncappedTarget` which is undesired according to the definition of the `uncappedTarget` recommendation.

adrianmoisey replied:
I'm not quite sure I understand your response, but my comments weren't very clear. 
I'm not suggesting changing to Pod-level maximum flags.

1. Does it make sense to move these flags into `pkg/recommender/logic/recommender.go`, so that all the max/min flags live near each other in the code base.
2. Does it make sense to rename the `max-allowed-cpu` and `max-allowed-memory` flags (that are being introduced in this PR) to `container-recommendation-max-memory-mb` and `container-recommendation-min-cpu-millicores`, similar to what you described in https://github.com/kubernetes/autoscaler/issues/7147#issuecomment-2501042262

Sorry for misunderstanding, I hope that clears it up?

ialidzhikov replied:
Yes, it makes sense to prefix the flags with `--container-recommendation` to make the purpose of the flags clear.
I addressed the suggestion in https://github.com/kubernetes/autoscaler/pull/7560/commits/7e427ae770fbe13b53c80e188c5c9d55a4cd2ec3.

adrianmoisey commented on vertical-pod-autoscaler/pkg/utils/vpa/capping.go:
```diff
@@ -176,23 +176,43 @@ func applyVPAPolicy(recommendation apiv1.ResourceList, policy *vpa_types.Contain
 		for resourceName, recommended := range recommendation {
-			cappedToMin, _ := maybeCapToPolicyMin(recommended, resourceName, containerPolicy)
-			recommendation[resourceName] = cappedToMin
-			cappedToMax, _ := maybeCapToPolicyMax(cappedToMin, resourceName, containerPolicy)
-			recommendation[resourceName] = cappedToMax
+			// containerPolicy can be nil (user does not have to configure it).
+			if containerPolicy != nil {
+				cappedToMin, _ := maybeCapToPolicyMin(recommended, resourceName, containerPolicy)
+				recommendation[resourceName] = cappedToMin
+
+				maxAllowed := containerPolicy.MaxAllowed
+				if globalMaxAllowed != nil {
+					if maxAllowed == nil {
+						maxAllowed = globalMaxAllowed
+					} else {
+						// Set resources from the global maxAllowed if the VPA maxAllowed is missing them.
+						for resourceName, quantity := range globalMaxAllowed {
+							if _, ok := maxAllowed[resourceName]; !ok {
+								maxAllowed[resourceName] = quantity
+							}
+						}
+					}
+				}
+
+				cappedToMax, _ := maybeCapToMax(cappedToMin, resourceName, maxAllowed)
+				recommendation[resourceName] = cappedToMax
+			} else {
+				if globalMaxAllowed != nil {
+					cappedToMax, _ := maybeCapToMax(recommended, resourceName, globalMaxAllowed)
+					recommendation[resourceName] = cappedToMax
+				}
+			}
 		}
 	}
```
Would this be a little more simpler to read?
```
	process := func(recommendation apiv1.ResourceList) {
		for resourceName, recommended := range recommendation {
			var maxAllowed apiv1.ResourceList
                        // containerPolicy can be nil (user does not have to configure it).
			if containerPolicy != nil {
				cappedToMin, _ := maybeCapToPolicyMin(recommended, resourceName, containerPolicy)
				recommendation[resourceName] = cappedToMin
				maxAllowed = containerPolicy.MaxAllowed
			}

			if globalMaxAllowed != nil {
				if maxAllowed == nil {
					maxAllowed = globalMaxAllowed
				} else {
					for resourceName, quantity := range globalMaxAllowed {
						if _, ok := maxAllowed[resourceName]; !ok {
							maxAllowed[resourceName] = quantity
						}
					}
				}
			}

			cappedToMax, _ := maybeCapToMax(recommendation[resourceName], resourceName, maxAllowed)
			recommendation[resourceName] = cappedToMax
		}
	}
```

ialidzhikov replied:
Agree. I addressed the suggestion in https://github.com/kubernetes/autoscaler/pull/7560/commits/04e33401fe93d83320414e87f30570cdc24586de.

adrianmoisey commented on vertical-pod-autoscaler/pkg/recommender/main.go:
```diff
@@ -215,8 +221,16 @@ func run(healthCheck *metrics.HealthCheck, commonFlag *common.CommonFlags) {
 		postProcessors = append(postProcessors, &routines.IntegerCPUPostProcessor{})
 	}
 
+	var globalMaxAllowed apiv1.ResourceList
+	if !maxAllowedCPU.Quantity.IsZero() {
+		setGlobalMaxAllowed(&globalMaxAllowed, apiv1.ResourceCPU, maxAllowedCPU.Quantity)
+	}
+	if !maxAllowedMemory.Quantity.IsZero() {
+		setGlobalMaxAllowed(&globalMaxAllowed, apiv1.ResourceMemory, maxAllowedMemory.Quantity)
```
Should there be some logic to ensure that the global max is greater than the global min?
It may be a bit strange since one is per container and the other per pod

adrianmoisey replied:
> However, I am not sure how to add such validation when the global min allowed flags are on Pod-level and global max allowed flags are on container-level.

Yeah, I agree. I can't figure out a way to make the validation work.

> In [#7147 (comment)](https://github.com/kubernetes/autoscaler/issues/7147#issuecomment-2501042262) I suggested to deprecate the global Pod-level min allowed flags because:

I don't think deprecation is necessary yet. Pod level resources [may be in the VPA's future](https://github.com/kubernetes/autoscaler/issues/7571), so those flags may be used for Pod level resources.

adrianmoisey commented on vertical-pod-autoscaler/docs/examples.md:
```diff
@@ -108,3 +109,16 @@ These options cannot be used together and are mutually exclusive.
 Please use this option with caution as it may be possible to break Pod creation if there is a failure with the VPA.
 Using it in conjunction with `--ignored-vpa-object-namespaces=kube-system` or `--vpa-object-namespace` to reduce risk.
+
+### Specifying global maximum allowed resources to prevent pods from being unschedulable
```
What do you think about moving this section move to the "features.md" page?

ialidzhikov replied:
There are many sections in examples.md which actually describe features of VPA (Starting multiple recommenders, Custom memory bump-up after OOMKill, Using CPU management with static policy, Controlling eviction behavior based on scaling direction and resource, etc.). I don't think the newly introduced section is different from the existing section. 
IMO, examples and features are overlapping conceptually. If you describe a feature, you usually also add example(s) of how the feature can be used. `examples.md` and `features.md` should be merged IMO. This is out-of-scope of the existing PR.

adrianmoisey replied:
> IMO, examples and features are overlapping conceptually. If you describe a feature, you usually also add example(s) of how the feature can be used. `examples.md` and `features.md` should be merged IMO. This is out-of-scope of the existing PR.

I agree with you here. The features page is new, and I want to start moving the examples across to the features page, with more description about that feature. 

I thought it would be nice for the "global max" feature to be added to features from the start, since it's a better fit there, and will require work to move at a later stage.

adrianmoisey replied:
(but it's up to you though, the documentation needs a lot of work in general)

ialidzhikov replied:
I would prefer to keep it consistent with the existing doc. In another PR all the examples doc can be reworked as features sections.

adrianmoisey commented:
Unsure if you want to do this, but an e2e test for the feature would be **great**!
If it's not something you want do, I'm happy to try figure it out in a different PR

ialidzhikov commented:
> Unsure if you want to do this, but an e2e test for the feature would be **great**!

Do you have a suggestion for the e2e test? It seems that we deploy vpa to a kind cluster and then create/update/delete VPA objects and finally assert their state. It is relatively easy to test a field in the VPA spec. But I don't find example of e2e tests for such global flags/options.
The only option I see is to deploy VPA initially with the global max allowed flags. I am not sure if this is desired or not. It would also affect other recommender tests. Or do you suggest to patch the vpa-recommender deployment as part of the e2e test?

ialidzhikov commented:
```
ERROR: (gcloud.compute.networks.create) Could not fetch resource:
 - <!DOCTYPE html>
<html lang=en>
  <meta charset=utf-8>
  <meta name=viewport content="initial-scale=1, minimum-scale=1, width=device-width">
  <title>Error 502 (Server Error)!!1</title>
  <style>
    *{margin:0;padding:0}html,code{font:15px/22px arial,sans-serif}html{background:#fff;color:#222;padding:15px}body{margin:7% auto 0;max-width:390px;min-height:180px;padding:30px 0 15px}* > body{background:url(//www.google.com/images/errors/robot.png) 100% 5px no-repeat;padding-right:205px}p{margin:11px 0 22px;overflow:hidden}ins{color:#777;text-decoration:none}a img{border:0}@media screen and (max-width:772px){body{background:none;margin-top:0;max-width:none;padding-right:0}}#logo{background:url(//www.google.com/images/branding/googlelogo/1x/googlelogo_color_150x54dp.png) no-repeat;margin-left:-5px}@media only screen and (min-resolution:192dpi){#logo{background:url(//www.google.com/images/branding/googlelogo/2x/googlelogo_color_150x54dp.png) no-repeat 0% 0%/100% 100%;-moz-border-image:url(//www.google.com/images/branding/googlelogo/2x/googlelogo_color_150x54dp.png) 0}}@media only screen and (-webkit-min-device-pixel-ratio:2){#logo{background:url(//www.google.com/images/branding/googlelogo/2x/googlelogo_color_150x54dp.png) no-repeat;-webkit-background-size:100% 100%}}#logo{display:inline-block;height:54px;width:150px}
  </style>
  <a href=//www.google.com/><span id=logo aria-label=Google></span></a>
  <p><b>502.</b> <ins>That’s an error.</ins>
  <p>The server encountered a temporary error and could not complete your request.<p>Please try again in 30 seconds.  <ins>That’s all we know.</ins>
```

---

/test pull-kubernetes-e2e-autoscaling-vpa-full

adrianmoisey commented:
> > Unsure if you want to do this, but an e2e test for the feature would be **great**!
> 
> Do you have a suggestion for the e2e test? It seems that we deploy vpa components to a kind cluster and then create/update/delete VPA objects and finally assert their state. It is relatively easy to test a field in the VPA spec. But I don't find example of e2e tests for such global flags/options. The only option I see is to deploy VPA initially with the global max allowed flags. I am not sure if this is desired or not. It would also affect other recommender tests. Or do you suggest to patch the vpa-recommender deployment as part of the e2e test?

It seems that at the moment we don't have a similar test to base yours off, so I think one would need to be written to deploy the recommender with a global max, and ensure that it doesn't go above that max.

It's going to be annoying to write such a test, I think, but I think the test will help us in the long run.

What do you think?

adrianmoisey commented:
I'm going to give this a
/lgtm

I like the idea of an e2e test, but will let an approver decide if it's needed or not

raywainman commented on vertical-pod-autoscaler/pkg/recommender/main.go:
```diff
@@ -116,6 +119,9 @@ const (
 )
 
 func main() {
+	flag.Var(&maxAllowedCPU, "container-recommendation-max-allowed-cpu", "Maximum amount of CPU that will be recommended for a container. VerticalPodAutoscaler-level maximum allowed takes precedence over the global maximum allowed.")
```
Let's move these into an `init()` function in this file?

https://go.dev/doc/effective_go#init

It is guaranteed to run before `main()`.

raywainman commented on vertical-pod-autoscaler/pkg/recommender/main.go:
```diff
@@ -307,3 +321,11 @@ func run(healthCheck *metrics.HealthCheck, commonFlag *common.CommonFlags) {
 	}
 }
+
+func setGlobalMaxAllowed(globalMaxAllowed *apiv1.ResourceList, key apiv1.ResourceName, value resource.Quantity) {
```
What are your thoughts on converting this function into an "initializer" for the globalMaxAllowed object?

In `main()` you would do:

`globalMaxAllowed := initGlobalMaxAllowed()`

And then in this function:

```
func initGlobalMaxAllowed() apiv1.ResourceList {
  result := make(apiv1.ResourceList)
  if !maxAllowedCPU.Quantity.IsZero() {
    result[apiv1.ResourceCPU] = maxAllowedCPU.Quantity
  }
  ... same for memory ...
  return result
}
```

This way you avoid playing around with pointers and the function is fully self contained.

raywainman commented on vertical-pod-autoscaler/pkg/utils/vpa/capping.go:
```diff
@@ -184,22 +184,39 @@ func applyVPAPolicy(recommendation apiv1.ResourceList, policy *vpa_types.Contain
+				maxAllowed = containerPolicy.MaxAllowed
+			}
+
+			if globalMaxAllowed != nil {
```
to simplify, we can assume this is never nil since you are initializing it up the stack (even if the user is not setting any global max values)

Also `ResourceList` is a map[] under the hood which Go just treats as an "empty map" when it is nil so we don't need to worry about panics here.

(this will remove one level of checking here)

raywainman commented on vertical-pod-autoscaler/pkg/utils/vpa/capping.go:
```diff
@@ -184,22 +184,39 @@ func applyVPAPolicy(recommendation apiv1.ResourceList, policy *vpa_types.Contain
+				} else {
+					// Set resources from the global maxAllowed if the VPA maxAllowed is missing them.
+					for resourceName, quantity := range globalMaxAllowed {
+						if _, ok := maxAllowed[resourceName]; !ok {
```
Let's add a comment here that we only override this if the user did not explicitly set a maximum in their container policy in the VPA.

ialidzhikov replied:
There is already https://github.com/kubernetes/autoscaler/blob/04e33401fe93d83320414e87f30570cdc24586de/vertical-pod-autoscaler/pkg/utils/vpa/capping.go#L210 which is more or less stating the same thing.
Let me know if you have suggestions for improvements.

raywainman commented on vertical-pod-autoscaler/pkg/utils/vpa/capping_test.go:
```diff
@@ -283,67 +283,256 @@ func TestApply(t *testing.T) {
+			},
+		},
+		{
+			Name:              "resource policy has max allowed for cpu and global max allowed is set for cpu and memory",
```
what are your thoughts on having a test where only one of `cpu` or `memory` is set for global max allowed?

raywainman reviewed (commented):
Thanks for this!!

ialidzhikov commented:
Hi @raywainman ,

I addressed your review feedback in https://github.com/kubernetes/autoscaler/pull/7560/commits/caa47f9d644a1bec7599d6ea9dd5134ea0330d0d.

adrianmoisey commented:
/lgtm

adrianmoisey commented:
/milestone vertical-pod-autoscaler-1.4.0

raywainman commented on vertical-pod-autoscaler/pkg/utils/vpa/capping.go:
```diff
@@ -184,22 +184,37 @@ func applyVPAPolicy(recommendation apiv1.ResourceList, policy *vpa_types.Contain
-			cappedToMin, _ := maybeCapToPolicyMin(recommended, resourceName, containerPolicy)
-			recommendation[resourceName] = cappedToMin
-			cappedToMax, _ := maybeCapToPolicyMax(cappedToMin, resourceName, containerPolicy)
+			var maxAllowed apiv1.ResourceList
```
Can we compute `maxAllowed` outside the for-loop? I think that simplifies the loop a little bit...

```
// Container Policy always takes precedence over the global max allowed value.
var maxAllowed apiv1.ResourceList
if containerPolicy != nil {
	maxAllowed = containerPolicy.MaxAllowed
} else if globalMaxAllowed != nil {
	maxAllowed = globalMaxAllowed
}
...

for resourceName, recommended := range recommendation {
	if containerPolicy != nil {
		cappedToMin, _ := maybeCapToPolicyMin(recommended, resourceName, containerPolicy)
		recommendation[resourceName] = cappedToMin
	}
	if maxAllowed != nil {
		cappedToMax, _ := maybeCapToMax(recommendation[resourceName], resourceName, maxAllowed)
		recommendation[resourceName] = cappedToMax
	}
    }
}
```

What do you think?

ialidzhikov replied:
Hi @raywainman ,

In general, pretty valid comment. It totally makes sense to compute the minAllowed and maxAllowed values outside of the for loop.

In your suggestion, there is no merge between the containerPolicy max allowed and the global max allowed.

The equivalent suggestion with the proper merge would be:
```golang
var minAllowed apiv1.ResourceList
if containerPolicy != nil {
	minAllowed = containerPolicy.MinAllowed
}

var maxAllowed apiv1.ResourceList
if containerPolicy != nil {
	maxAllowed = containerPolicy.MaxAllowed
}
if maxAllowed == nil {
	maxAllowed = globalMaxAllowed
} else {
	// Set resources from the global max allowed if the VPA max allowed is missing them.
	// Container policy max allowed always takes precedence over the global max allowed value.
	for resourceName, quantity := range globalMaxAllowed {
		if _, ok := maxAllowed[resourceName]; !ok {
			maxAllowed[resourceName] = quantity
		}
	}
}

process := func(recommendation apiv1.ResourceList) {
	for resourceName, _ := range recommendation {
		if minAllowed != nil {
			cappedToMin, _ := maybeCapToMin(recommendation[resourceName], resourceName, minAllowed)
			recommendation[resourceName] = cappedToMin
		}
		if maxAllowed != nil {
			cappedToMax, _ := maybeCapToMax(recommendation[resourceName], resourceName, maxAllowed)
			recommendation[resourceName] = cappedToMax
		}
	}
}
```

Tomorrow, I will perform a final test and if everything is fine, I will push a commit that incorporates the feedback.
Thank you very much! 

Very nice suggestion, I like it!

ialidzhikov replied:
@raywainman , I addressed your suggestion in https://github.com/kubernetes/autoscaler/pull/7560/commits/055affac85a28e559d8fe1e8cb3d09e19e697180.

ialidzhikov replied:
@plkokanov , I think you are right. In https://github.com/kubernetes/autoscaler/blob/055affac85a28e559d8fe1e8cb3d09e19e697180/vertical-pod-autoscaler/pkg/utils/vpa/capping.go#L201-L204 I assign the `maxAllowed` var to `containerPolicy.MaxAllowed`. Later on, in https://github.com/kubernetes/autoscaler/blob/055affac85a28e559d8fe1e8cb3d09e19e697180/vertical-pod-autoscaler/pkg/utils/vpa/capping.go#L208-L213, I modify the `maxAllowed` var with content from the globalMaxAllowed, if needed.

Let me double check this.

ialidzhikov replied:
A snippet that demonstrates the problem that @plkokanov discovered: https://go.dev/play/p/TGgl4FA8-5w

Example output:
```
maxAllowed=map[cpu:200m memory:300Mi], 0xc0000a0120
containerPolicy.MaxAllowed=map[cpu:200m memory:300Mi], 0xc0000a0120
```

`containerPolicy.MaxAllowed` is modified in unexpected manner.

ialidzhikov replied:
@plkokanov , I addressed your suggestion in https://github.com/kubernetes/autoscaler/pull/7560/commits/bab77cd33848aafe1e3bd4655c98c962c7a8ee88.

raywainman replied:
Good call on the deep copy. This LGTM now, thank you!

raywainman reviewed (commented):
I sincerely apologize for the long lag time here, please take a look at one suggestion below. I think otherwise this is good to go :)

Thanks for all of this!!

adrianmoisey commented:
/lgtm

plkokanov reviewed (commented):
/lgtm

raywainman commented:
/approve
